docs/flows.md

```diff
@@ -0,0 +1,90 @@
+# Common Authentication Data Flows
+
+## Authorization Creation (ie. Login)
+1. Client authenticates without JWT. (ie. email/password, oauth, etc)
+2. Invalidate any existing authorizations and refresh tokens for that user/client combo.
+3. Create a new authorization record in the database with
+  - id (unique)
+  - entityId (unique)
+  - clientId (unique)
+  - permissions (maybe if they get default permissions)
+   
+  _A authorization can have any number of verifying ids/values (ie. we could also add an organizationId, MFA pin, etc. into the mix)_
+
+4. Create an access token (JWT) with the verification values/ids:
+  - authorizationId
+  - entityId (verification values)
+  - clientId
+  
+  **NOTE:** This could now actually have a long TTL
+5. Create a refresh token with a very long TTL
+  - refreshId
+  - entityId (verification values)
+  - clientId
+6. Send back the access token (JWT) and refresh token (JWT) to the client
+
+## Access Token Verification
+The access token (JWT) is passed via an `Authorization` header **or** is attached to the connected socket. These are requests made to anything but the `/authentication` service.
+
+1. Client sends valid JWT access token or it's attached to the connected socket
+2. Server verifies JWT (in our special auth middleware) to ensure it is not:
+  - expired (then it removes or flags authorization as revoked/expired)
+  - malformed
+  - tampered with
+3. Server decodes the token payload and passes it down the auth middleware chain
+4. Server looks up authorization record by the `authorizationId` and passes it down the auth middleware chain (populateAuthorization)
+5. Your own custom auth middleware validates the authorization
+  - if valid then proceed to the service before hooks or any other auth middleware.
+  - if invalid set `data.authenticated = false` and move on to next middleware
+6. Service `before` hook, `isAuthenticated` checks to see if the request is authenticated.
+    - If it is proceed to next hook or service method
+    - If it isn't it rejects with a `NotAuthenticated` error
+7. Service `before` hook, `isPermitted` checks to see if the request is authenticated.
+    - If it is proceed to next hook or service method
+    - If it isn't it rejects with a `Forbidden` error
+8. Service method is called and returns through after hooks
```

Comment:
Steps 6 and 7 above seem too much alike? afaik, 401 Not authenticated is meant for when the client has not yet presented any credentials (so pre-login), whereas 403 Forbidden is meant for clients that did show credentials (so post-login) but are simply not allowed access. So I would expect the one hook to check whether the client is authenticated, whereas the other should check whether he is authorized to perform the action.